src/neml2/models/Model.cxx

```diff
@@ -32,45 +32,254 @@ ParameterSet
 {
-  LabeledVector out(in.batch_size(), output());
-  LabeledMatrix dout_din(out, in);
-  set_value(in, out, &dout_din);
+  LabeledVector out(in.batch_size(), {&output()}, in.options());
+  LabeledMatrix dout_din(in.batch_size(), {&output(), &in.axis()}, in.options());
+
+  if (_AD_1st_deriv)
+  {
+    // Set requires_grad to true if not already
+    bool req_grad = in.tensor().requires_grad();
+    in.tensor().requires_grad_();
+    // Evaluate the model value
+    set_value(in, &out);
+    // Loop over rows to retrieve the derivatives
+    for (TorchSize i = 0; i < out.tensor().base_sizes()[0]; i++)
+    {
+      BatchTensor<1> grad_outputs = torch::zeros_like(out.tensor());
+      grad_outputs.index_put_({torch::indexing::Ellipsis, i}, 1.0);
+      try
+      {
+        auto jac_row =
+            torch::autograd::grad({out.tensor()}, {in.tensor()}, {grad_outputs}, true, true)[0];
+        dout_din.tensor().base_index_put({i, torch::indexing::Slice()}, jac_row);
+      }
+      catch (c10::Error &)
+      {
+        // This is aggravating: libTorch throws if the derivative is zero... but why?!
+      }
+    }
+    in.tensor().requires_grad_(req_grad);
+  }
+  else
+    set_value(in, &out, &dout_din);
+
   return {out, dout_din};
 }
 
+std::tuple<LabeledMatrix, LabeledTensor3D>
+Model::dvalue_and_d2value(const LabeledVector & in) const
+{
+  if (_AD_1st_deriv || _AD_2nd_deriv)
+    return {std::get<1>(value_and_dvalue_and_d2value(in)),
+            std::get<2>(value_and_dvalue_and_d2value(in))};
+
+  LabeledMatrix dout_din(in.batch_size(), {&output(), &in.axis()}, in.options());
+  LabeledTensor3D d2out_din2(in.batch_size(), {&output(), &in.axis(), &in.axis()}, in.options());
+  set_value(in, nullptr, &dout_din, &d2out_din2);
+  return {dout_din, d2out_din2};
+}
+
+std::tuple<LabeledVector, LabeledMatrix, LabeledTensor3D>
+Model::value_and_dvalue_and_d2value(const LabeledVector & in) const
+{
+  LabeledVector out(in.batch_size(), {&output()}, in.options());
+  LabeledMatrix dout_din(in.batch_size(), {&output(), &in.axis()}, in.options());
+  LabeledTensor3D d2out_din2(in.batch_size(), {&output(), &in.axis(), &in.axis()}, in.options());
+
+  if (_AD_2nd_deriv)
+  {
+    // Set requires_grad to true if not already
+    bool req_grad = in.tensor().requires_grad();
+    in.tensor().requires_grad_();
+    if (_AD_1st_deriv)
+    {
+      set_value(in, &out);
+      // Loop over rows to retrieve the derivatives
+      for (TorchSize i = 0; i < out.tensor().base_sizes()[0]; i++)
+      {
+        BatchTensor<1> grad_outputs = torch::zeros_like(out.tensor());
+        grad_outputs.index_put_({torch::indexing::Ellipsis, i}, 1.0);
+        try
+        {
+          auto jac_row =
+              torch::autograd::grad({out.tensor()}, {in.tensor()}, {grad_outputs}, true, true)[0];
+          dout_din.tensor().base_index_put({i, torch::indexing::Slice()}, jac_row);
+        }
+        catch (c10::Error &)
```

Comment:
Fix this either by:

1. Figuring out the right torch options to grad...
2. Detecting the issue with logic in the loop versus catching here
3. Worst case: check error string...

Reply:
Fixed -- see above.